man/wscp.1
.Dd October 17, 2013
.Dt WSCP 1
.Os
.Sh NAME
.Nm wscp
.Nd execute commands on multiple remote machines
.Sh SYNOPSIS
.Nm wshc
.Op Fl -port Ar port
.Op Fl u | -username Ar username
.Op Fl p | -password
.Op Fl t | -threads Ar threads
.Op Fl l | -location Ar destination
.Fl h | -hosts | f | -file | r | -range Ar hosts
.Ar files
.Sh DESCRIPTION
.Nm
logs in to multiple hosts and executes commands on them, in parallel. It's
suitable for executing commands on large numbers of hosts and collating their
results in a format that's easy to read.
.Ss Arguments
.Bl -tag -width u
.It Fl -port Ar port
Use a different port for ssh. By default, 22 is used.
.It Fl u | -username Ar username
Use a different username for SSH. By default, the username of the current
user is used.
.It Fl p | -password
Prompt for the SSH password. If this argument is not provided, then
.Nm
will assume that you are using pubkeys or an agent.
.It Fl t | -threads Ar threads
Spawn off
.Ar threads
number of threads. If this flag is not given, then
.Nm
will execute sequentially.
.It Fl l | -location Ar destination
Destination on the remote host where
.Nm
drops its file.
.Nm
expects the remote directory to exist and will fail if it doesn't.
.El
.Ss Host selection arguments
.Bl -tag -width u
.It Fl h | -hosts Ar hosts
A comma separated list of hosts to ssh into.
.It Fl f | -file Ar filename
A file to read a newline delimited set of hosts from. If
.Ar filename
is executable,
.Nm
will execute the file and use the output as the list of hosts.
.It Fl r | -range Ar range_query
If
.Nm
is built with range support,
.Ar range_query
will be resolved with range
.El
.Sh EXIT STATUS
.Ex -std
.Sh SEE ALSO
.Xr wshc 1
.Xr wshd 1
.Xr wsh-add-hostkeys 1
.Sh AUTHORS
.An William Orr Aq Mt will@worrbase.com
.Sh SECURITY CONSIDERATIONS
.Pp
.Nm
prompts for your passwords and stores them in memory. A page of memory is
.Xr mlock 2
and the password is kept in memory. With some knowledge of the
source and some time, it is possible to fetch the password from memory.
.Pp
As soon as
.Nm
is done with the password, it is scrubbed from memory with
.Xr memset_s 3
.Pp
The input buffer is directed to the
.Xr mlock 2
page of memory with
.Xr setvbuf 3
and also scrubbed with
.Xr memset_s 3
after use.


util/src/wscp.c
#include <glib.h>
#include <stdlib.h>
#include <string.h>

#include "client.h"
#include "expansion.h"
#include "log.h"
#include "ssh.h"
#ifndef HAVE_MEMSET_S
extern int memset_s(void* v, size_t smax, int c, size_t n);
#endif

static gint port = 22;
static gchar* username = NULL;
static gboolean ask_password = FALSE;
static gint threads = 0;
static gchar* location = NULL;

// Host selection variables
static gchar* hosts_arg = NULL;
static gchar* file_arg = NULL;
static gchar* range = NULL;

static void* passwd_mem;

static GOptionEntry entries[] = {
	{ "port", 0, 0, G_OPTION_ARG_INT, &port, "Port to use, if not 22", NULL },
	{ "username", 'u', 0, G_OPTION_ARG_STRING, &username, "SSH username", NULL },
	{ "password", 'p', 0, G_OPTION_ARG_NONE, &ask_password, "Prompt for SSH password", NULL },
	{ "threads", 't', 0, G_OPTION_ARG_INT, &threads, "Number of threads to spawn", NULL },
	{ "location", 'l', 0, G_OPTION_ARG_STRING, &location, "Location on remote hosts to drop files", NULL },

	// Host selection
	{ "hosts", 'h', 0, G_OPTION_ARG_STRING, &hosts_arg, "Comma separated list of hosts to ssh into", NULL },
	{ "file", 'f', 0, G_OPTION_ARG_STRING, &file_arg, "Filename to read hosts from", NULL },
#ifdef RANGE
	{ "range", 'r', 0, G_OPTION_ARG_STRING, &range, "Range query for hostname expansion", NULL },
#endif
	{ NULL },
};

static gboolean valid_arguments(gchar** mesg) {
	if ((hosts_arg && (file_arg || range)) || (file_arg && range)) {
		*mesg = g_strdup("Use one of -h, -r or -f\n");
		return FALSE;
	}

	if (!(hosts_arg || file_arg || range)) {
		*mesg = g_strdup("Use one of -h, -r or -f\n");
		return FALSE;
	}

	return TRUE;
}

typedef struct {
	gchar** files;
	gchar* host;
	gchar* user;
	gchar* pass;
	gchar* location;
	gint num_files;
	gint port;
} wshc_scp_file_args;

static gint scp_file(const wshc_scp_file_args* args) {
	GError *err = NULL;
	wsh_ssh_session_t session = {
		.session = NULL,
		.channel = NULL,
		.hostname = args->host,
		.username = args->user,
		.password = args->pass,
		.scp = NULL,
		.port = args->port,
	};

	if (session.password == NULL)
		session.auth_type = WSH_SSH_AUTH_PUBKEY;
	else
		session.auth_type = WSH_SSH_AUTH_PASSWORD;

	if (wsh_ssh_host(&session, &err)) {
		g_printerr("%s\n", err->message);
		g_error_free(err);
		return EXIT_FAILURE;
	}

	if (wsh_verify_host_key(&session, FALSE, FALSE, &err)) {
		g_printerr("%s\n", err->message);
		g_error_free(err);
		return EXIT_FAILURE;
	}

	if (wsh_ssh_authenticate(&session, &err)) {
		g_printerr("%s\n", err->message);
		g_error_free(err);
		return EXIT_FAILURE;
	}

	if (wsh_ssh_scp_init(&session, args->location)) {
		g_printerr("Error initializing scp");
		return EXIT_FAILURE;
	}

	for (gint i = 0; i < args->num_files; i++) {
		if (wsh_ssh_scp_file(&session, args->files[i], FALSE, &err)) {
			g_printerr("%s\n", err->message);
			g_error_free(err);
			err = NULL;
		}
	}

	wsh_ssh_scp_cleanup(&session);
	wsh_ssh_disconnect(&session);

	return EXIT_SUCCESS;
}

gint main(gint argc, gchar** argv) {
	GError* err = NULL;
	GOptionContext* context;
	gint ret = EXIT_SUCCESS;
	gchar** hosts = NULL;
	gsize num_hosts = 0;
	gchar* password = NULL;

#if ! GLIB_CHECK_VERSION( 2, 32, 0 )
	g_thread_init(NULL);
#endif

	wsh_init_logger(WSH_LOGGER_CLIENT);
	wsh_ssh_init();

	context = g_option_context_new("[FILENAMES...] - scp file to multiple machines at once");
	g_option_context_add_main_entries(context, entries, NULL);
	if (! g_option_context_parse(context, &argc, &argv, &err)) {
		g_printerr("Option parsing failed: %s\n", err->message);
		g_error_free(err);
		return EXIT_FAILURE;
	}

	if (argc == 1) {
		g_printerr("ERROR: Must provide a host and a file\n\n");
		g_printerr("%s", g_option_context_get_help(context, FALSE, NULL));
		g_option_context_free(context);
		return EXIT_FAILURE;
	}

	gchar* msg = NULL;
	if (! valid_arguments(&msg)) {
		g_printerr("%s\n", msg);
		g_free(msg);

		g_printerr("%s", g_option_context_get_help(context, FALSE, NULL));
		g_option_context_free(context);
		return EXIT_FAILURE;
	}

	g_option_context_free(context);

	// We're done with option validation
	if (username == NULL)
		username = g_strdup(g_get_user_name());

	if (ask_password) {
		if ((ret = wsh_client_lock_password_pages(passwd_mem))) {
			g_printerr("%s\n", strerror(ret));
			return ret;
		}

		password = ((gchar*)passwd_mem) + (WSH_MAX_PASSWORD_LEN * 0);
		if ((ret = wsh_client_getpass(password, WSH_MAX_PASSWORD_LEN, "SSH password: ", passwd_mem))) {
			g_printerr("%s\n", strerror(ret));
			return ret;
		}

		if (! password) return EXIT_FAILURE;
	}

	if (!location)
		location = g_strdup("~");

	if (file_arg) {
		if (wsh_exp_filename(&hosts, &num_hosts, file_arg, &err)) {
			g_printerr("%s\n", err->message);
			g_error_free(err);
			return EXIT_FAILURE;
		}
	}

	if (hosts_arg) {
		hosts = g_strsplit(hosts_arg, ",", 0);
		num_hosts = g_strv_length(hosts);
	}

#ifdef RANGE
	if (range) {
		if (wsh_exp_range(&hosts, &num_hosts, range, &err)) {
			g_printerr("%s\n", err->message);
			g_error_free(err);
			return EXIT_FAILURE;
		}
	}
#endif

	argv++; argc--;
	for (gint i = 1; i < argc; i++) {
		// scp will catch this, but I'd like to catch it before
		// we start even trying to scp files
		if (!g_file_test(argv[i], G_FILE_TEST_EXISTS)) {
			g_printerr("ERROR: %s doesn't exist\n", argv[i]);
			ret = EXIT_FAILURE;
			goto bad;
		}
	}

	if (threads <= 0) {
		wshc_scp_file_args args;
		args.port = port;
		args.user = username;
		args.pass = password;
		args.files = argv;
		args.num_files = argc;
		args.location = location;

		for (gsize i = 0; i < num_hosts; i++) {
			args.host = hosts[i];
			scp_file(&args);
		}
	} else {
		GThreadPool* gtp;
		if ((gtp = g_thread_pool_new((GFunc)scp_file, NULL, threads, TRUE, &err)) == NULL) {
			g_printerr("%s\n", err->message);
			g_error_free(err);
			return EXIT_FAILURE;
		}

		wshc_scp_file_args args[num_hosts];

		for (gsize i = 0; i < num_hosts; i++) {
			args[i].host = hosts[i];
			args[i].port = port;
			args[i].user = username;
			args[i].pass = password;
			args[i].files = argv;
			args[i].num_files = argc;
			args[i].location = location;

			g_thread_pool_push(gtp, &args[i], NULL);
		}

		g_thread_pool_free(gtp, FALSE, TRUE);
	}

	if (password) {
		memset_s(password, WSH_MAX_PASSWORD_LEN, 0, strlen(password));
		wsh_client_unlock_password_pages(passwd_mem);
	}

bad:
	wsh_ssh_cleanup();
	wsh_exit_logger();

	if (password) g_free(password);
	password = NULL;
	if (range) g_free(range);
	range = NULL;
	if (hosts_arg) g_free(hosts_arg);

	g_free(username);
	username = NULL;

	return ret;
}

